File: msc_edgar/roles.py
"""Officer role normalization and ranking (brief §3d).

A single normalized-substring matcher maps any free-text role to a seniority
priority (lower = more senior). Keys are matched **longest-first** so
"President and CEO" resolves to CEO (1), never President (2) — the founder-CEO
under-ranking trap called out in the brief.
"""

from __future__ import annotations

import re

from config import OFFICER_ROLE_PRIORITY, ROLE_PRIORITY_UNKNOWN

# Role keywords precomputed longest-first for substring matching.
_ROLE_KEYS = sorted(OFFICER_ROLE_PRIORITY, key=len, reverse=True)

# Any of these keyword stems makes a text cell look like a role/title.
ROLE_HINTS = (
    "chief", "officer", "president", "ceo", "cfo", "coo", "cto", "director",
    "chairman", "chairperson", "chair", "founder", "secretary", "treasurer",
    "general counsel", "managing member", "manager", "partner", "principal",
    "vice president", "vp", "trustee", "executive",
)


def normalize_role(role: str) -> str:
    """Lowercase, expand ``&`` to ``and``, strip punctuation, collapse spaces."""
    if not role:
        return ""
    s = role.lower().replace("&", " and ")
    s = re.sub(r"[^a-z0-9 ]+", " ", s)
    return re.sub(r"\s+", " ", s).strip()


def role_priority(role: str) -> int:
    """Map a role string to its seniority priority (lower = more senior)."""
    norm = normalize_role(role)
    if not norm:
        return ROLE_PRIORITY_UNKNOWN
    for key in _ROLE_KEYS:          # longest key first
        if key in norm:
            return OFFICER_ROLE_PRIORITY[key]
    return ROLE_PRIORITY_UNKNOWN


def looks_like_role(text: str) -> bool:
    """True if a text cell plausibly describes a title/role."""
    norm = normalize_role(text)
    return any(h in norm for h in ROLE_HINTS)

File: worker/src/roles.ts
// Officer role normalization and ranking (ported from msc_edgar/roles.py).
// A normalized-substring matcher maps free-text roles to a seniority priority
// (lower = more senior). Keys are matched LONGEST-FIRST so "President and CEO"
// resolves to CEO (1), not President (2).

export const OFFICER_ROLE_PRIORITY: Record<string, number> = {
  "chief executive officer": 1, "president and ceo": 1, "president & ceo": 1, "ceo": 1,
  "president": 2,
  "co-founder": 3, "founder": 3,
  "chairman": 4, "chairperson": 4, "chair": 4,
  "chief financial officer": 5, "cfo": 5,
  "chief operating officer": 6, "coo": 6,
  "general counsel": 7, "chief legal officer": 7,
  "secretary": 8,
  "executive officer": 9,
  "director": 10,
};
export const ROLE_PRIORITY_UNKNOWN = 99;
export const TOP_N_OFFICERS = 3;

// Longest key first for substring matching.
const ROLE_KEYS = Object.keys(OFFICER_ROLE_PRIORITY).sort((a, b) => b.length - a.length);

const ROLE_HINTS = [
  "chief", "officer", "president", "ceo", "cfo", "coo", "cto", "director",
  "chairman", "chairperson", "chair", "founder", "secretary", "treasurer",
  "general counsel", "managing member", "manager", "partner", "principal",
  "vice president", "vp", "trustee", "executive",
];

export function normalizeRole(role: string): string {
  if (!role) return "";
  let s = role.toLowerCase().replace(/&/g, " and ");
  s = s.replace(/[^a-z0-9 ]+/g, " ");
  return s.replace(/\s+/g, " ").trim();
}

export function rolePriority(role: string): number {
  const norm = normalizeRole(role);
  if (!norm) return ROLE_PRIORITY_UNKNOWN;
  for (const key of ROLE_KEYS) {
    if (norm.includes(key)) return OFFICER_ROLE_PRIORITY[key];
  }
  return ROLE_PRIORITY_UNKNOWN;
}

export function looksLikeRole(text: string): boolean {
  const norm = normalizeRole(text);
  return ROLE_HINTS.some((h) => norm.includes(h));
}
